frontend/js/src/brainzplayer/AppleMusicPlayer.tsx

```diff
@@ -0,0 +1,372 @@
+    );
+  };
+
+  static getURLFromListen(listen: Listen | JSPFTrack): string | undefined {
+    return _get(listen, "track_metadata.additional_info.apple_music_id");
+  }
+
+  public name = "Apple Music";
+  public domainName = "music.apple.com";
+  public icon = faApple;
+  // Saving the developer token outside of React state , we do not need it for any rendering purposes
+  // and it simplifies some of the closure issues we've had with old tokens.
+  private readonly developerToken: string = "";
+  private readonly musicUserToken: string = "";
+
+  private readonly _boundOnPlaybackStateChange: (
+    event: MusicKit.PlayerPlaybackState
+  ) => void;
+
+  private readonly _boundOnPlaybackTimeChange: (
+    event: MusicKit.PlayerPlaybackTime
+  ) => void;
+
+  private readonly _boundOnPlaybackDurationChange: (
+    event: MusicKit.PlayerDurationTime
+  ) => void;
+
+  private readonly _boundOnNowPlayingItemChange: (
+    event: MusicKit.NowPlayingItem
+  ) => void;
+
+  appleMusicPlayer?: AppleMusicPlayerType;
+
+  constructor(props: AppleMusicPlayerProps) {
+    super(props);
+    this.state = {
+      durationMs: 0,
+      progressMs: 0,
+    };
+
+    this.developerToken = props.appleMusicUser?.developer_token || "";
+    this.musicUserToken = props.appleMusicUser?.music_user_token || "";
+
+    // Do an initial check of whether the user wants to link Apple Music before loading the SDK library
+    if (AppleMusicPlayer.hasPermissions(props.appleMusicUser)) {
+      window.addEventListener("musickitloaded", this.connectAppleMusicPlayer);
+      loadScriptAsync(
+        document,
+        "https://js-cdn.music.apple.com/musickit/v3/musickit.js"
+      );
+    } else {
+      this.handleAccountError();
+    }
+
+    this._boundOnPlaybackStateChange = this.onPlaybackStateChange.bind(this);
+    this._boundOnPlaybackTimeChange = this.onPlaybackTimeChange.bind(this);
+    this._boundOnPlaybackDurationChange = this.onPlaybackDurationChange.bind(
+      this
+    );
+    this._boundOnNowPlayingItemChange = this.onNowPlayingItemChange.bind(this);
+  }
+
+  componentDidUpdate(prevProps: DataSourceProps) {
+    const { show } = this.props;
+    if (prevProps.show && !show) {
+      this.stopAndClear();
+    }
+  }
+
+  componentWillUnmount(): void {
+    this.disconnectAppleMusicPlayer();
+  }
+
+  playAppleMusicId = async (
+    appleMusicId: string,
+    retryCount = 0
+  ): Promise<void> => {
+    const { handleError } = this.props;
+    if (retryCount > 5) {
+      handleError("Could not play AppleMusic track", "Playback error");
+      return;
+    }
+    if (!this.appleMusicPlayer || !this.appleMusicPlayer.isAuthorized) {
+      await this.connectAppleMusicPlayer();
+      return;
+    }
+    try {
+      await this.appleMusicPlayer.setQueue({
+        song: appleMusicId,
+        startPlaying: true,
+      });
+    } catch (error) {
+      handleError(error.message, "Error playing on Apple Music");
```

Comment:
Something should happen after catching this error. handleError only shows the message in a toast notification. 
Not sure what should happen though, it should probably depend on the type of error if we can diferentiate that.
Should it retry? should it call `handleAccountError`? `onTrackNotFound` if the song can't be found (unlikely I guess)?
I would again point to the Spotify player as an example for this: 
https://github.com/metabrainz/listenbrainz-server/blob/93c392f9756ff5b688610ecc460bf604afcddc1a/frontend/js/src/brainzplayer/SpotifyPlayer.tsx#L259-L286